### ISSUE TITLE
Create library-specific versions of logits processors

### ISSUE THREAD

**Issue by RobinPicard** [dottxt-ai/outlines#1445]
First suggested in #1444:

> Not for this PR, but I think we'll have to break down the logits processor into library-specific ones. On top of this we'll soon have methods in outlines-core to mask on-device that see optimized for torch, MLX and NumPy. We can already open an issue for that.

The justification for this proposed change is that libraries providing local models (transformers, mlxlm, lamma.cpp...) use different object types and shapes to represent logits.
